webapp/src/components/tutorialValidators.tsx

```diff
@@ -0,0 +1,171 @@
+
+    constructor(properties: { [index: string]: string }) {
+        super(properties);
+        this.useHintHighlight = properties["useanswerkeyhighlight"]?.toLowerCase() === "true";
+    }
+
+    async executeInternal(options: CodeValidationExecuteOptions): Promise<CodeValidationResult> {
+        let missingBlocks: string[] = [];
+        let disabledBlocks: string[] = [];
+        const {parent, tutorialOptions} = options;
+        const stepInfo = tutorialOptions.tutorialStepInfo
+            ? tutorialOptions.tutorialStepInfo[tutorialOptions.tutorialStep]
+            : null;
+        if (!stepInfo) return { isValid: true, hint: "" };
+
+        // Currently, this validator only supports useHintHighlight, so this flag/check is redundant,
+        // but there are plans to support manual specification of blocks in the future.
+        // As such, it felt prudent to still require this flag so tutorial authors don't have to go back and add it later on.
+        if (this.useHintHighlight) {
+            const userBlocks = Blockly.getMainWorkspace()?.getAllBlocks(false /* ordered */);
+            const userBlocksEnabledByType = new Map<string, boolean>(); // Key = type, value = enabled
+            userBlocks.forEach(b => userBlocksEnabledByType.set(b.type, userBlocksEnabledByType.get(b.type) || b.isEnabled()));
+
+            const allHighlightedBlocks = await getTutorialHighlightedBlocks(tutorialOptions, stepInfo);
+            if(!allHighlightedBlocks) {
+                return defaultResult;
+            }
+
+            const stepHash = getTutorialStepHash(tutorialOptions);
+            const stepHighlights = allHighlightedBlocks[stepHash];
+            const highlightedBlockKeys = stepHighlights ? Object.keys(stepHighlights) : [];
+
+            for (let i: number = 0; i < highlightedBlockKeys.length; i++) {
+                let tutorialBlockKey = highlightedBlockKeys[i];
+                const isEnabled = userBlocksEnabledByType.get(tutorialBlockKey);
+                if (isEnabled === undefined) {
+                    // user did not use a specific block
+                    missingBlocks.push(tutorialBlockKey);
+                } else if(!isEnabled) {
+                    disabledBlocks.push(tutorialBlockKey);
+                }
+            }
+        }
+
+        let isValid = true;
+        let errorDescription: string;
+
+        if (missingBlocks.length > 0 && disabledBlocks.length > 0) {
+            isValid = false;
+            errorDescription = lf("Make sure you see blocks that look like this and that they're connected to the rest of your code.")
+        } else if (missingBlocks.length > 0) {
+            isValid = false;
+            errorDescription = lf("Make sure you see blocks that look like this in your workspace.");
+        } else if (disabledBlocks.length > 0) {
+            isValid = false;
+            errorDescription = lf("Make sure your blocks are connected to the rest of your code like this.");
```

Comment:
@kiki-lee Would love to get your thoughts on the three different error descriptions here. One is for when there are both missing and detached blocks, one is when there's only missing blocks, and one is when there's only detached blocks.

Reply:
Checked with Kiki offline.